crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -46,15 +51,17 @@ where
     S: Eq + Hash + std::fmt::Debug + Send + Sync + Clone + 'static,
 {
     fn get(&self, key: &CacheKey<S>) -> Option<CacheEntry> {
-        self.0.get(key)
+        // LruMap's get() method requires a mutable reference since it updates the LRU order
+        // So we need to use a write lock here
+        self.0.write().get(key).cloned()
```

Comment:
imo this is also mine, but then we can simply use parking_lot::Mutex

Reply:
about the `RwLock` vs `Mutex` - you're right that since we're using `write()` for both reads and writes due to LruMap's implementation, a Mutex would be more appropriate. I'll make that change.